frontend/src/metabase/visualizations/echarts/cartesian/option/series.ts

```diff
@@ -74,16 +74,45 @@ export const getBarLabelLayout =
 
 export function getDataLabelFormatter(
   seriesModel: SeriesModel,
+  dataset: ChartDataset,
   settings: ComputedVisualizationSettings,
   labelDataKey: DataKey,
   renderingContext: RenderingContext,
   formattingOptions: OptionsType = {},
 ) {
+  const shouldRenderCompact = (() => {
+    if (settings["graph.label_value_formatting"] === "compact") {
+      return true;
+    }
+    if (settings["graph.label_value_formatting"] === "full") {
+      return false;
+    }
+    // for "auto" we use compact if it shortens avg label length by >3 chars
+    const getAvgLength = (compact: boolean) => {
+      const lengths = dataset.map(datum => {
+        const value = datum[labelDataKey];
+        return renderingContext.formatValue(value, {
+          ...(settings.column?.(seriesModel.column) ?? {}),
+          jsx: false,
+          compact: compact,
+          ...formattingOptions,
+        }).length;
+      });
+
+      return (
+        lengths.reduce((sum: number, length: number) => sum + length, 0) /
+        lengths.length
+      );
+    };
+
+    return getAvgLength(true) + 3 < getAvgLength(false);
```

Comment:
Otherwise, compute the average data label lengths for `full` and `compact` formatting.

If the `compact` formatting average length is more than 3 less than the `full` formatting length, use `compact` formatting. Otherwise, use `full`.